Add forward AD layout check for storage numel

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68631
* #68629

This PR:
- Adds the check that the storage numel of the base and tangent tensors are the same. This is to support the case when as_strided reveals elements that aren't indexable by the input tensor.
- Skips the check when batched tensors are involved, because using as_strided to reveal elements that not indexable by the input tensor is already not allowed vmap.
- Adds tests for the above two cases, as well as an edge case regarding conj bit (what about neg bit?)

For functorch:
- we need to copy the batching rule implemented here

Differential Revision: [D32899678](https://our.internmc.facebook.com/intern/diff/D32899678)

soulitzer committed:
Add forward AD layout check for storage numel

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD layout check for storage numel"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add forward AD layout check for storage numel"


- Instead of a native function to compare two tensors, maybe a native function that just retrieves the storage numel. 
- Using the stride of the first batch dim for batched tensors, seems like a reasonable replacement for storage numel.


[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add forward AD layout check for storage numel"


- Instead of a native function to compare two tensors, maybe a native function that just retrieves the storage numel. 
- Using the stride of the first batch dim for batched tensors, seems like a reasonable replacement for storage numel.


[ghstack-poisoned]

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -563,13 +563,11 @@ Tensor _new_zeros_with_same_feature_meta_batching_rule(
+  auto physical_view = at::MultiBatchVmapTransform::logicalToPhysical(self);
+  auto num_batch_dims = physical_view.numBatchDims();
+  const auto& physical_tensor = physical_view.tensor();
+  return physical_tensor.strides()[num_batch_dims - 1];
```
This would only be valid if this is conrtiguous and the batch dims are first no?

soulitzer replied:
Ah yes, we'd need to add the check enforcing this.

`_new_zeros_with_same_feature_meta`  makes sure the the storage of the logical tensor as a whole is tiled contiguously, the logical tensor itself need not be contiguous

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2432,9 +2432,5 @@
   other: non_differentiable
   output_differentiability: [False]
 
-- name: _has_same_storage_numel(Tensor self, Tensor other) -> bool
```
The new function still needs an entry here no?

soulitzer replied:
The number of differentiable outputs would be zero in this case because the only output is a non-Tensor, so I think that the behavior in terms of codegen would be the same?

albanD replied:
Ok

soulitzer replied:
Here it is!
```
int64_t _storage_numel(c10::DispatchKeySet ks, const at::Tensor & self) {
  auto& self_ = unpack(self, "self", 0);
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::_storage_numel(ks & c10::after_autograd_keyset, self_);
  })();
  auto result = _tmp;
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  return result;
}
```

soulitzer committed:
Update on "Add forward AD layout check for storage numel"


- Instead of a native function to compare two tensors, maybe a native function that just retrieves the storage numel. 
- Using the stride of the first batch dim for batched tensors, seems like a reasonable replacement for storage numel.


[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -11453,6 +11453,8 @@ def ref_pairwise_distance(input1, input2):
            check_batched_gradgrad=False,
            supports_forward_ad=True,
+           # See https://github.com/pytorch/pytorch/issues/66357
+           check_batched_forward_grad=False,
```
This is weird... still investigating.

soulitzer replied:
I figured out why this was the case. This failing started happening after adding the check that batch dimensions are at the front. 

See https://github.com/pytorch/pytorch/pull/68631#discussion_r761365724

soulitzer committed:
Update on "Add forward AD layout check for storage numel"


- Instead of a native function to compare two tensors, maybe a native function that just retrieves the storage numel. 
- Using the stride of the first batch dim for batched tensors, seems like a reasonable replacement for storage numel.


[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD layout check for storage numel"


- Instead of a native function to compare two tensors, maybe a native function that just retrieves the storage numel. 
- Using the stride of the first batch dim for batched tensors, seems like a reasonable replacement for storage numel.


[ghstack-poisoned]

soulitzer commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -472,7 +472,15 @@ Tensor view_as_complex_batching_rule(const Tensor& self) {
+    std::string func_name) {
+  if (std::find(physical_sizes.begin(), physical_sizes.end(), 0) != physical_sizes.end()) {
+    // The tensor has zero numel
+    return;
```
Not sure if this is the right approach because I'm not clear on exactly what cases this check guards against. Does this check still make sense to do if the numel of the tensor is zero?

I added this to address the zero-numel case where `_new_zeros_with_same_feature_meta` prepends zeros as the strides if the storage numel of the tensor is zero. 

Relaxing this check also seemed to address some failures:
- several in https://github.com/pytorch/pytorch/issues/66357.
- batched gradgrad tests for linalg ops (they were failing previously on empty inputs)

Sidenote: there were other batched gradgrad (and even some batched grad) tests in linalg and fft that I seem to pass even without this fix. Having xfail instead of skip would be much nicer.

cc @zou3519

soulitzer replied:
I initially understood this check to be "are the batch dimensions at the front". But I think there is a distinction between that and "are the batch dimensions at the front in layout" which is what this function actually seems to check.

Maybe this is not the check we need in `_new_zeros_with_same_meta` and `_storage_numel`?

zou3519 replied:
Why do you need to modify this function?

Since we're almost done with getting vmap + forward-mode AD to compose in functorch, I want to stop supporting the vmap in core because it's too much unnecessary work

soulitzer replied:
Good point. Maybe we wouldn't need to do it here, but do you think it makes sense to modify it in functorch though?

zou3519 replied:
I thought about this some more -- if the tensor has zero numel then any batch dims are "trivially at the front in memory layout" because there is no storage. So yes it would make sense to modify it in functorch

zou3519 commented on torch/csrc/autograd/autograd_meta.cpp:
```diff
@@ -85,6 +85,9 @@ namespace {
         return false;
       }
     }
+    if (base._storage_numel() != other._storage_numel()) {
```
By the way, there's a huge performance problem with `_new_zeros_with_same_feature_meta` that arises with the invariant that `base._storage_numel() == other._storage_numel()`

Let's say we have `x: Tensor[5], y: Tensor[3]` and we construct a dual tensor `make_dual(primal=x[0], tangent=BatchedTensor(y))`.

When we set `BatchedTensor(y)` to be the tangent, we require it to have the same storage numel as primal. So we end up creating coercing the tangent into a tensor of shape [3, 5].

If the user never actually uses as_strided on their dual tensor or does in-place operations, then they are paying for memory that they didn't ask for.

How important is it that we add this check? I want to say that forward-mode AD does not need this check during composition with vmap.

soulitzer replied:
Thanks for the example, it does seem like it would be an easy way for users to shoot themselves in the foot.

Looks like (A) is already unsupported by vmap as well (in core and in functorch)? In that case I would agree that forward AD does not need this check in composition with vmap.

```
from functorch import vmap
from torch._vmap_internals import _vmap

x = torch.rand(10, 2).select(1, 0)

# Both fail
vmap(lambda x: x.as_strided([1], [1], 1))(x)
_vmap(lambda x: x.as_strided([1], [1], 1))(x)
```

Also filed this issue regarding the error message: https://github.com/pytorch/functorch/issues/308

zou3519 replied:
> Looks like (A) is already unsupported by vmap as well (in core and in functorch)?

Nice catch! I forgot we had those checks

soulitzer committed:
Update on "Add forward AD layout check for storage numel"


- Instead of a native function to compare two tensors, maybe a native function that just retrieves the storage numel. 
- Using the stride of the first batch dim for batched tensors, seems like a reasonable replacement for storage numel.


[ghstack-poisoned]

albanD commented on aten/src/ATen/ConjugateFallback.cpp:
```diff
@@ -31,6 +31,9 @@ TORCH_LIBRARY_IMPL(aten, Conjugate, m) {
   m.impl("resolve_neg", torch::CppFunction::makeFallthrough());
 
+  // See test_layout_check_for_primal_with_conj_bit in test_autograd.py
+  m.impl("_has_same_storage_numel", torch::CppFunction::makeFallthrough());
```
If you have this one, you most likely also want it for neg version.
And you want a special impl for zero tensor as well? That returns true like the batched version?

soulitzer replied:
> If you have this one, you most likely also want it for neg version.

Yup.

> And you want a special impl for zero tensor as well? That returns true like the batched version?

Also a good point. Zero tensors aren't landed yet though right?

albanD replied:
> Zero tensors aren't landed yet though right?

It was but maybe got reverted?

albanD replied:
It was reverted indeed.

albanD commented on torch/csrc/autograd/autograd_meta.cpp:
```diff
@@ -85,6 +85,9 @@ namespace {
         return false;
       }
     }
+    if (!at::_has_same_storage_numel(base, other)) {
```
This function used to be "cheap" as it was never going through a full dispatch. Is there a measurable impact to this call?

soulitzer replied:
Benchmark to compare:
```
Before: 2.26 us, 1638366 instr
After : 2.44 us, 1686272 ins
```

<details>
<summary>
code
</summary>

```python
setup="""
import torch.autograd.forward_ad as fwAD
import torch

primal = torch.rand(10, 10)
tangent = torch.rand(10, 10)
try:
    torch._C._exit_dual_level(level=0)
except:
    pass
torch._C._enter_dual_level()
"""

stmt="""
torch._VF._make_dual(primal, tangent, level=0)
"""

timer = Timer(stmt, setup)

print(timer.timeit(10000))
print(timer.collect_callgrind(100))
```

</details>

From measuring `make_dual` there actually seems to be a measurable difference of around ~300 ns (~0.3 us) overhead, which seems pretty expected for dispatcher from what I remember from benchmarking the autograd-not-implemented kernel.

The relative slowdown is probably even larger if we call directly into `_set_fw_grad`, but just to imagine what the impact of this would be, if a model computes _set_fw_grad 1000 times, it would take 0.3ms longer to run.

albanD replied:
Sounds good!

albanD reviewed (commented):
Sounds mostly good. CI needs to be fixed though.

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899678).

soulitzer committed:
Update on "Add forward AD layout check for storage numel"


This PR:
- Adds the check that the storage numel of the base and tangent tensors are the same. This is to support the case when as_strided reveals elements that aren't indexable by the input tensor.
- Skips the check when batched tensors are involved, because using as_strided to reveal elements that not indexable by the input tensor is already not allowed vmap.
- Adds tests for the above two cases, as well as an edge case regarding conj bit (what about neg bit?)

For functorch:
- we need to copy the batching rule implemented here

Differential Revision: [D32899678](https://our.internmc.facebook.com/intern/diff/D32899678)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899678).

soulitzer committed:
Update on "Add forward AD layout check for storage numel"


This PR:
- Adds the check that the storage numel of the base and tangent tensors are the same. This is to support the case when as_strided reveals elements that aren't indexable by the input tensor.
- Skips the check when batched tensors are involved, because using as_strided to reveal elements that not indexable by the input tensor is already not allowed vmap.
- Adds tests for the above two cases, as well as an edge case regarding conj bit (what about neg bit?)

For functorch:
- we need to copy the batching rule implemented here

Differential Revision: [D32899678](https://our.internmc.facebook.com/intern/diff/D32899678)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -132,6 +132,17 @@
+# always returns true because vmapped as_strided does not support accessing
+# storage locations not indexable by the input tensor.
+# See the note above for more information.
+- func: _has_same_storage_numel(Tensor self, Tensor other) -> bool
```
In the long term, maybe there should be two different modes for forward-mode AD? One that supports tensor subclasses, and one that supports the fast-path tensor code. If there are two different modes like that then we probably won't need a `_has_same_storage_numel` operator.

But the approach in this PR makes sense from a short-to-mid-term perspective. It's a bit unfortunate that we have to register all these additional implementations for various dispatch keys for this new operator.

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899678).

soulitzer commented:
The other failure should be unrelated, but `test_ops` is failing because we need to wait for https://github.com/pytorch/pytorch/pull/69558 to land.

albanD reviewed (approved):
ok!

soulitzer committed:
Update on "Add forward AD layout check for storage numel"


This PR:
- Adds the check that the storage numel of the base and tangent tensors are the same. This is to support the case when as_strided reveals elements that aren't indexable by the input tensor.
- Skips the check when batched tensors are involved, because using as_strided to reveal elements that not indexable by the input tensor is already not allowed vmap.
- Adds tests for the above two cases, as well as an edge case regarding conj bit (what about neg bit?)

For functorch:
- we need to copy the batching rule implemented here

Differential Revision: [D32899678](https://our.internmc.facebook.com/intern/diff/D32899678)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32899678).
